service/storage_service.cc

```diff
@@ -557,9 +557,13 @@ future<> storage_service::topology_state_load() {
     // will be up to date and reachable at the time of restart.
     const auto tmptr = get_token_metadata_ptr();
     for (const auto& e: tmptr->get_all_endpoints()) {
+        if (is_me(e)) {
+            continue;
+        }
         const auto ep = tmptr->get_endpoint_for_host_id(e);
-        if (!is_me(e) && !_gossiper.get_endpoint_state_ptr(ep)) {
-            co_await _gossiper.add_saved_endpoint(ep);
+        auto permit = co_await _gossiper.lock_endpoint(ep, gms::null_permit_id);
```

Comment:
I don't understand something about `add_saved_endpoint` call in `storage_service::join_cluster` - we are load ing IPs/tokens from `system.peers` and constructing `token_metadata` from them. We call `add_saved_endpoint` in the loop, by it uses the old version of `token_metadata`, not the one we'are constructing. This means we probably never go under the `if (host_id) {` in it and never fill `HOST_ID` and `TOKENS`.

Reply:
there is another place where `add_saved_endpoint` is called - `storage_service::join_token_ring`:
```
co_await _gossiper.reset_endpoint_state_map();
for (auto ep : loaded_endpoints) {
    co_await _gossiper.add_saved_endpoint(ep);
}
```
so it probably works my this means. By why then we need the `add_saved_endpoint` call in `join_cluster`?

Reply:
In join_cluster we get the host_id from system.peers.
This is more evident in https://github.com/scylladb/scylladb/pull/15788/commits/a4b158d1d5d055d198514bb1af4fb1c38c6a9952

In `join_token_ring`, we call `add_saved_endpoints` after `do_shadow_round` which clears the endpoint states.
This is orthogonal to this PR, but I think it'd be possible to keep the loaded state in the gossiper (using `loaded_endpoint_state` to be introduced in `https://github.com/scylladb/scylladb/pull/15788`, and loaded automatically in `gossiper::reset_endpoint_state_map` so the reset endpoint state map will contain the state loaded from system.peers.

Reply:
> In join_cluster we get the host_id from system.peers.
> This is more evident in [a4b158d](https://github.com/scylladb/scylladb/commit/a4b158d1d5d055d198514bb1af4fb1c38c6a9952)

This is not "more evident", this is actually changed behaviour — IICU, without this commit we don't see `host_id` and `tokens` for ep in `join_cluster/add_saved_endpoint` call.

> This is orthogonal to this PR, but I think it'd be possible to keep the loaded state in the gossiper (using `loaded_endpoint_state` to be introduced in `https://github.com/scylladb/scylladb/pull/15788`, and apply the loaded endpoint state automatically in `gossiper::reset_endpoint_state_map` so the reset endpoint state map will contain the state loaded from system.peers.

This sounds cleaner.

Reply:
@bhalevy why do we need to call add_saved_endpoint more than once? (join_cluster, join_token_ring). It seems this needs to be cleared first.

Reply:
I explained it in some other comment in this PR IIRC.
The other call is needed after shadow round that resets the endpoint states map.
This is not needed to be cleared first because this all happens before gossip starts so the endpoint state has generation=0 anyhow so updating it more than once in this state via add_saved_endpoint is not a problem.

Reply:
So why does the shadow round reset the endpoint states map?

Reply:
I discussed the resetting of the endpoint state with @asias and maybe he can explain it better than me.

In any case, I'd like to avoid too intrusive changes in the scope of this PR to make it feasible for backport.
Let's consider design changes for improvement as follow up.

https://github.com/scylladb/scylladb/pull/16432#issuecomment-1869523583 is one such direction.

As for loading the state once, see https://github.com/scylladb/scylladb/pull/16432#discussion_r1431495073 earlier in this thread.
It'd be possible with https://github.com/scylladb/scylladb/pull/16432#issuecomment-1869523583, when `topology_state_load` will stop abusing `add_saved_endpoint` to inject application states, but rather it would that using an official api similar to `add_local_application_state`.

Reply:
But please, let's keep in mind that this is a point fix that should be backportable and hold our breath with fixing the whole world around it.

Reply:
So the patch has no test, the impact on the clients (is anyone suffering from this?) is unclear,  and is working around a problem that only @asias knows why it exists :(

@gusev-p do you understand why we have to add endpoints twice? @tgrabiec perhaps you?

Reply:
@bhalevy why should it be backportable?

Reply:
> @bhalevy why should it be backportable?

The context is https://github.com/scylladb/scylla-enterprise/issues/3518 and https://github.com/scylladb/scylladb/pull/15788

> So the patch has no test,

It is covered by the existing topology tests.

> the impact on the clients (is anyone suffering from this?) is unclear, and is working around a problem that only @asias knows why it exists :(

The was stumbled upon in review.  See https://github.com/scylladb/scylladb/pull/15788#discussion_r1428382450

You keep returning to "why we have to add endpoints twice?", I presume in `join_cluster` and `join_token_ring`, but this is really not problematic WRT to the endpoint state generation, since both calls happen before gossip starts gossiping - i.e. the endpoint state will have generation=0 in both cases.

The problem happens via `topology_state_load` that uses `add_saved_endpoint` to inject the node's state, yet it may lose the race to gossiper and discover in `add_saved_endpoint` that the state was already loaded, after the latter acquires the endpoint_lock.

To prevent that (and yes, we don't want to add the state more than once), we want `topology_state_load` to acquire the endpoint lock for checking if the endpoint_state already exists.
And to do that, we want https://github.com/scylladb/scylladb/pull/16507 to release the lock around waiting in error injection so that `topology_state_load` can grab it.

I hope that clarifies the background for this PR and #16507.

Reply:
1) I believe your response is contradictory, sorry. If the behavior is covered by the existing tests, then how is 3518 possibl?
2) I checked 3518 and the customer did not follow our topology operations best practices. The behavior is known and is addressed by the topology changes on Raft. Moreover, it's part of the test plan for this.

So what's the point of fixing this in gossip mode in 6.0? If you wish to satisfy the customer, perhaps direct these patches straight to the enterprise release branch.

Reply:
> 1. I believe your response is contradictory, sorry. If the behavior is covered by the existing tests, then how is 3518 possible?

I may have misunderstood your question then.
I'm talking about this PR and the access to the endpoint state via add_saved_endpoint that's covered, not about scylladb/scylla-enterprise#3518.
For the latter, I added dtest reproducers in https://github.com/scylladb/scylla-dtest/pull/3685
that naturally were added after it was hit.

> 2. I checked 3518 and the customer did not follow our topology operations best practices. The behavior is known and is addressed by the topology changes on Raft. Moreover, it's part of the test plan for this feature.

The fact that it's addressed by raft topology changes doesn't mean we don't want to fix that in older releases.
It will take considerable time until consistent topology changes reach the field in production, and FE (and me) want some solution in place until then.

> So what's the point of fixing this in gossip mode in 6.0? If you wish to satisfy the customer, perhaps direct these patches straight to the enterprise release branch.

No, this is not enterprise only code. The fixes must go through master.

And it's beyond satisfying the customer.
We have a hole in add_saved_endpoint that doesn't populate DC and RACK and the changes to support it are relatively simple.
Trying to over engineer it is like killing a spider with a sledgehammer :)